compiler/src/dmd/expressionsem.d

```diff
@@ -1155,6 +1155,43 @@ L1:
     return e1;
 }
 
+/*
+ * Check whether `outerFunc` and `calledFunc` have the same `this`.
+ * If `calledFunc` is the member of a base class of the class that contains
+ * `outerFunc` we consider that they have the same this.
+ *
+ * Returns:
+ *  `true` if outerFunc and calledFunc may use the same `this` pointer.
+ * `false` otherwise.
+ */
+private bool haveSameThis(FuncDeclaration outerFunc, FuncDeclaration calledFunc)
```

Comment:
This function doesn't look right. The logic in the function is surprisingly complex (many branches). I would expect `haveSameThis` to be commutative and take two arbitrary functions, but the names `outerFunc` and `calledFunc` suggest there are restrictions to the parameters which aren't documented.

Reply:
Initially, I thought the fix is easy. I thought it would suffice to make sure that the this of the function containing the call and the this of the function being called are the same, however that does not cut it for derived member functions. Later on, I noticed that for nested functions there is some extra engineering done to get the right this. That's when I ended up with this function. I suspect that there are other places where this function might be used.

I don't think it can be made any simpler than that, but I am confident that what is in there at this point is correct.